BacktraceScreen.c

```diff
@@ -0,0 +1,384 @@
+/*
+htop - BacktraceScreen.h
+(C) 2024 htop dev team
+Released under the GNU GPLv2+, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "config.h" // IWYU pragma: keep
+
+#include "BacktraceScreen.h"
+
+#if defined(HAVE_BACKTRACE)
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/wait.h>
+
+#include "CRT.h"
+#include "FunctionBar.h"
+#include "ListItem.h"
+#include "Macros.h"
+#include "Object.h"
+#include "Panel.h"
+#include "Platform.h"
+#include "Process.h"
+#include "ProvideCurses.h"
+#include "RichString.h"
+#include "Settings.h"
+#include "Vector.h"
+#include "XUtils.h"
+
+
+#define BACKTRACE_FRAME_HEADER_INDEX -1
+
+#define BACKTRACE_PANEL_OPTIONS_NAME_DEMANGLE "Demangle"
+#define BACKTRACE_PANEL_OPTIONS_NAME_RAW "Raw"
+
+#define BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_FULL "Full Path"
+#define BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_BASENAME "Basename"
+
+#define BACTRACE_PANEL_HEADER_NUMBER_FRAME "#"
+#define BACTRACE_PANEL_HEADER_ADDRESS "ADDRESS"
+#define BACTRACE_PANEL_HEADER_NAME "NAME"
+#define BACTRACE_PANEL_HEADER_PATH "PATH"
+
+#define MAX_ADDR_32 (0xffffffff)
+#define MAX_HEX_ADDR_STR_LEN_32 (strlen("0x") + 8)
+#define MAX_HEX_ADDR_STR_LEN_64 (strlen("0x") + 16)
+
+static const char* const BacktraceScreenFunctions[] = {
+   "Refresh",
+#if defined(HAVE_DEMANGLING)
+   BACKTRACE_PANEL_OPTIONS_NAME_RAW,
+#endif
+   BACKTRACE_PANEL_OPTIONS_OBJECT_PATH_FULL,
+   "Done   ",
+   NULL
+};
+
+static const char* const BacktraceScreenKeys[] = {
+   "F1",
+#if defined(HAVE_DEMANGLING)
+   "F2",
+#endif
+   "F3",
+   "Esc",
+   NULL
+};
+
+static const int BacktraceScreenEvents[] = {
+   KEY_F(1),
+#if defined(HAVE_DEMANGLING)
+   KEY_F(2),
+#endif
+   KEY_F(3),
+   27,
+};
+
+static void BacktraceFrame_displayHeader(const BacktraceFrame* frame, RichString* out) {
+   const BacktracePanelPrintingHelper* printingHelper = &frame->backtracePanel->printingHelper;
+   assert(printingHelper);
```

Comment:
Redundant assert? (Since an `&` operator is used for retrieving address for `printingHelper`)

Reply:
Also, if at all you would check `assert(frame);`, which is better simply declared as `ATTR_NONNULL_ALL` because you want the `RichString*` to be valid too.

Reply:
Good Catch! Thank you.

Is it a good idea to put `ATTR_NONNULL` on every function? It will catch a lot of bugs.

Reply:
Yes and no. In the `htop` sources we don't always do it, but in strategic places you will find these compiler annotations. Normally the non-`NULL` is implied by skipping the `NULL` check or placing an `assert`. The attribute is only rarely used.

Reply:
Okay, I understand better. Thank you for the clarification @BenBE.